source/cl/source/extension/include/extension/intel_unified_shared_memory.h

```diff
@@ -254,6 +273,74 @@ class device_allocation_info final : public allocation_info {
   mux_buffer_t mux_buffer;
 };
 
+/// @brief Derived class for shared USM allocations
+class shared_allocation_info final : public allocation_info {
+ private:
+  /// @brief Private constructor, use the `create()` functions instead.
+  shared_allocation_info(const cl_context context, const cl_device_id device,
+                         const size_t size);
+
+  shared_allocation_info(const shared_allocation_info &) = delete;
+
+ public:
+  /// @brief Create a shared USM allocation
+  ///
+  /// @param[in] context Context the allocation will belong to.
+  /// @param[in] device Device associated with this shared allocation, may be
+  /// null to not associate it with a device.
+  /// @param[in] properties Properties bitfield encoding which properties to
+  /// enable.
+  /// @param[in] size Bytes to allocate.
+  /// @param[in] alignment Minimum alignment of allocation.
+  static cargo::expected<std::unique_ptr<shared_allocation_info>, cl_int>
+  create(cl_context context, cl_device_id device,
+         const cl_mem_properties_intel *properties, const size_t size,
+         cl_uint alignment);
+
+  /// @brief Destructor.
+  ~shared_allocation_info() override;
+
+  /// @brief Get device used when allocation was made, or nullptr if no device
+  /// was given
+  cl_device_id getDevice() const override { return device; }
+
+  /// @brief Allocates memory for the USM allocation and binds it to a Mux
+  /// buffer object if a device is given. Sets 'mux_memory' and 'mux_buffer'
```

Comment:
it says if a device is given - but the create requires a device?

Reply:
Create does not, `cl_device_id` is a typedefed pointer type.

Reply:
ah got you it does not require a device.